bigtable/tests/filters_integration_test.cc

```diff
@@ -0,0 +1,682 @@
+  // the separator and the successor of "/" is "0".
+  auto result =
+      ReadRows(client, table, row_key_prefix + "/", row_key_prefix + "0",
+               bigtable::Filter::CellsRowLimit(3));
+
+  std::map<std::string, int> actual;
+  for (auto const& c : result) {
+    auto ins = actual.emplace(static_cast<std::string>(c.row_key()), 0);
+    ins.first->second++;
+  }
+  std::map<std::string, int> expected{{row_key_prefix + "/one-cell", 1},
+                                      {row_key_prefix + "/two-cells", 2},
+                                      {row_key_prefix + "/many", 3},
+                                      {row_key_prefix + "/many-columns", 3},
+                                      {row_key_prefix + "/complex", 3}};
+
+  CheckEqualRowKeyCount(__func__, expected, actual);
+  std::cout << __func__ << " was successful" << std::endl;
+}
+
+void CheckCellsRowOffset(bigtable::Client& client, bigtable::Table& table,
+                         std::string const& row_key_prefix) {
+  CreateComplexRows(client, table, row_key_prefix);
+
+  // Search in the range [row_key_prefix, row_key_prefix + "0"), we used '/' as
+  // the separator and the successor of "/" is "0".
+  auto result =
+      ReadRows(client, table, row_key_prefix + "/", row_key_prefix + "0",
+               bigtable::Filter::CellsRowOffset(2));
+
+  std::map<std::string, int> actual;
+  for (auto const& c : result) {
+    auto ins = actual.emplace(static_cast<std::string>(c.row_key()), 0);
+    ins.first->second++;
+  }
+  std::map<std::string, int> expected{{row_key_prefix + "/many", 2},
+                                      {row_key_prefix + "/many-columns", 2},
+                                      {row_key_prefix + "/complex", 78}};
+
+  CheckEqualRowKeyCount(__func__, expected, actual);
+  std::cout << __func__ << " was successful" << std::endl;
+}
+
+void CreateManyRows(bigtable::Client& client, bigtable::Table& table,
+                    std::string const& prefix, int count) {
+  namespace bt = bigtable;
+  bt::BulkMutation bulk;
+  for (int row = 0; row != count; ++row) {
+    std::string row_key = prefix + "/" + std::to_string(row);
+    bulk.emplace_back(bt::SingleRowMutation(
+        row_key, {bt::SetCell("fam", "col", 4000, "foo")}));
+  }
+  table.BulkApply(std::move(bulk));
+}
+
+void CheckCellsRowSample(bigtable::Client& client, bigtable::Table& table,
+                         std::string const& row_key_prefix) {
+  constexpr int row_count = 20000;
+  CreateManyRows(client, table, row_key_prefix, row_count);
+
+  // We want to check that the sampling rate was "more or less" the prescribed
+  // value.  We use 5% as the allowed error, this is arbitrary.  If we wanted to
+  // get serious about testing the sampling rate we would do some statistics.
+  // We do not really need to, because we are testing the library, not the
+  // server. But for what is worth, the outline would be:
```

Comment:
s/for what is worth/for what it's worth/

Reply:
Fixed.